src/main/java/org/springframework/data/redis/cache/DefaultRedisCacheWriter.java

```diff
@@ -179,11 +179,14 @@ public void clean(String name, byte[] pattern) {
 					wasLocked = true;
 				}
 
-				byte[][] keys = Optional.ofNullable(connection.keys(pattern)).orElse(Collections.emptySet())
-						.toArray(new byte[0][]);
-
-				if (keys.length > 0) {
-					connection.del(keys);
+				List<byte[]> keys = new ArrayList<byte[]>();
+				Cursor<byte[]> cursor = connection.scan(new ScanOptions.ScanOptionsBuilder().match(new String(pattern)).build());
+				while (cursor.hasNext()) {
+					byte[] bytes = cursor.next();
+					keys.add(bytes);
+				}
+				if (!keys.isEmpty()) {
+					connection.del(keys.toArray(new byte[][]{}));
```

Comment:
As the complexity of DEL is O(N), if the number of keys is huge, DEL command could time out as well, right?
Maybe we could do it batch by batch as well?

I was thinking of using [UNLINK](https://redis.io/commands/unlink), but it's only supported in higher Redis version and it's a change of behavior (from sync to async)...